_posts/2022-08-04-calldata-tuple-reencoding-head-overflow-bug.md

```diff
@@ -0,0 +1,219 @@
+the unused space is guaranteed to be zeroed and remain clean after all high-level Solidity operations.
+In the affected cases the compiler would emit code that incorrectly cleaned the end of an array stored in the last component of a tuple,
+zeroing 32 bytes belonging to the first dynamically-sized component of the tuple.
+
+### Aggressive Array Cleanup
+
+When the input for the ABI encoder comes from calldata, large parts of it can simply be copied to the output.
+The encoder, however, must ensure that the unused spaces between elements are actually
+[cleaned](https://docs.soliditylang.org/en/latest/internals/variable_cleanup.html).
+
+Depending on the types, more or less work may be needed and the encoder has several code paths for handling them.
+One of them was designed for (possibly multi-dimensional) arrays of simple value types that fill the whole word,
+i.e. `uint` and `bytes32`, as well as byte arrays, i.e. `bytes` and `string`.
+In this case, due to tight packing of array elements, only the area past the array may need cleanup and that
+area is never larger than 32 bytes.
+
+The cleanup was being performed using a single `mstore()` instruction, which always writes exactly 32 bytes.
+The cleanup was only ever required for byte arrays but was not considered harmful for `uint` and `bytes32`,
+because of the order in which the encoder writes tuple components.
+The space after any given tuple component was assumed to be safe to write because it would eventually be
+overwritten with the data belonging to the next component.
+
+In order to explain why the assumption turned out to be incorrect, let us take a closer look at
+the layout of an ABI-encoded tuple.
+
+### Layout and Encoding Order of Statically- and Dynamically-sized Tuples
+
+[The ABI encoding of a tuple](https://docs.soliditylang.org/en/latest/abi-spec.html#formal-specification-of-the-encoding)
+consists of two areas - the statically-sized *head* and the dynamically-sized *tail*.
+Every statically-sized component is placed directly in the head area while for dynamically-sized ones, the head contains only the offset
+of the location within the tail where the data is stored.
+This layout ensures that component locations within the head are fixed and any of them can be accessed without the need to decode all
+the components that precede it.
+
+#### Statically-sized Example
+
+Let us consider the following example that encodes a tuple with no dynamically-sized components:
+```solidity
+struct S {
+    address x;
+    uint[3] y;
+}
+
+contract C {
+    function f(bool a, S calldata b, bytes32[2] calldata c)
+        public
+        returns (bool, S calldata, bytes32[2] calldata)
+    {
+        return (a, b, c);
+    }
+}
+```
+
+An example call to `C.f()` with the following input:
+
+```solidity
+f(true, S(0x1111111111222222222233333333334444444444, [uint(11), 12, 13]), [bytes32("a"), "b"])
+```
+
+should return the following values:
```

Comment:
```suggestion
should encode its return values as follows (which in this case is the same encoding that is used to pass the arguments to ``f``):
```
or something?
Might also be simpler to talk only about the argument encoding **or** the return value encoding, not some mixture...

Reply:
I tried to add that but I think that bringing attention to it actually seems a bit out of place (like, this detail is not really relevant to tuple encoding).

But yeah, the fact that the input and output are the same is also kinda distracting. So instead I reworked the examples. Now they do not return anything and I added a separate one for the bug. Check if that's better.